app/models/daos/PasswordInfoDAO.scala

```diff
@@ -0,0 +1,101 @@
+    dbPasswordInfo <- slickPasswordInfos if dbPasswordInfo.loginInfoId === dbLoginInfo.id
+  } yield dbPasswordInfo
+  
+  // Use subquery workaround instead of join to get authinfo because slick only supports selecting
+  // from a single table for update/delete queries (https://github.com/slick/slick/issues/684).
+  protected def passwordInfoSubQuery(loginInfo: LoginInfo) =
+    slickPasswordInfos.filter(_.loginInfoId in loginInfoQuery(loginInfo).map(_.id))
+
+  protected def addAction(loginInfo: LoginInfo, authInfo: PasswordInfo) =
+    loginInfoQuery(loginInfo).result.head.flatMap { dbLoginInfo =>
+      slickPasswordInfos +=
+        DBPasswordInfo(authInfo.hasher, authInfo.password, authInfo.salt, dbLoginInfo.id.get)
+    }.transactionally
+    
+  protected def updateAction(loginInfo: LoginInfo, authInfo: PasswordInfo) =
+    passwordInfoSubQuery(loginInfo).
+      map(dbPasswordInfo => (dbPasswordInfo.hasher, dbPasswordInfo.password, dbPasswordInfo.salt)).
+      update((authInfo.hasher, authInfo.password, authInfo.salt))
+  
+  /**
+   * Finds the auth info which is linked with the specified login info.
+   *
+   * @param loginInfo The linked login info.
+   * @return The retrieved auth info or None if no auth info could be retrieved for the given login info.
+   */
+  def find(loginInfo: LoginInfo): Future[Option[PasswordInfo]] = {
+    db.run(passwordInfoQuery(loginInfo).result.headOption).map { dbPasswordInfoOption =>
+      dbPasswordInfoOption.map(dbPasswordInfo => 
+        PasswordInfo(dbPasswordInfo.hasher, dbPasswordInfo.password, dbPasswordInfo.salt))
+    }
+  }
+
+  /**
+   * Adds new auth info for the given login info.
+   *
+   * @param loginInfo The login info for which the auth info should be added.
+   * @param authInfo The auth info to add.
+   * @return The added auth info.
+   */
+  def add(loginInfo: LoginInfo, authInfo: PasswordInfo): Future[PasswordInfo] =
+    db.run(addAction(loginInfo, authInfo)).map(_ => authInfo)
+
+  /**
+   * Updates the auth info for the given login info.
+   *
+   * @param loginInfo The login info for which the auth info should be updated.
+   * @param authInfo The auth info to update.
+   * @return The updated auth info.
+   */
+  def update(loginInfo: LoginInfo, authInfo: PasswordInfo): Future[PasswordInfo] = 
+    db.run(updateAction(loginInfo, authInfo)).map(_ => authInfo)
+
+  /**
+   * Saves the auth info for the given login info.
+   *
+   * This method either adds the auth info if it doesn't exists or it updates the auth info
+   * if it already exists.
+   *
+   * @param loginInfo The login info for which the auth info should be saved.
+   * @param authInfo The auth info to save.
+   * @return The saved auth info.
+   */
+  def save(loginInfo: LoginInfo, authInfo: PasswordInfo): Future[PasswordInfo] = {
+    val query = loginInfoQuery(loginInfo).joinLeft(slickPasswordInfos).on(_.id === _.loginInfoId)
+    val action = query.result.head.flatMap {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Usage of head on collections.](https://www.codacy.com/app/HiP-App/HiPCMS/file/2328382315/issues/source?bid=3144114&fileBranchId=3144125#l86)